### ISSUE TITLE
Memory leak when Cassandra connection issues

### ISSUE THREAD

**Issue by MichaelS11** [gocql/gocql#1139]
### What version of Cassandra are you using?
3.4.4

### What version of Gocql are you using?
From master, tested on a couple of versions from a few days ago to 5 days ago.

https://github.com/gocql/gocql/commit/4c9f01d25ad91a1fda9f93c75a47f121ef087c0f

### What did you do?
Connected to node when up, then brought node down.

### What did you expect to see?
It to fail and heap size not to grow

### What did you see instead?
Failed and heap size keep growing

---

pprof data:
```
1: 630784 [1: 630784] @ 0x409659 0x4097d7 0x40bb59 0x7b1fa7 0x7bac4b 0x7b8c3f 0x45b8d1
#	0x7b1fa6	github.com/gocql/gocql.(*Conn).exec+0x9a6		C:/Users/blah/src/github.com/gocql/gocql/conn.go:583
#	0x7bac4a	github.com/gocql/gocql.(*controlConn).writeFrame+0xba	C:/Users/blah/src/github.com/gocql/gocql/control.go:406
#	0x7b8c3e	github.com/gocql/gocql.(*controlConn).heartBeat+0x1ae	C:/Users/blah/src/github.com/gocql/gocql/control.go:76
```

I am new to gocql code. I was looking the below code link and noticed no call to releaseStream.
https://github.com/gocql/gocql/blob/master/conn.go#L618-L630

Yet in the following code link, releaseStream is called.
https://github.com/gocql/gocql/blob/master/conn.go#L658-L667

Is there a missing call to releaseStream if writeFrame has an error?

**Comment by MichaelS11** [comment]
In my test case, after about 28 hours, the heap did seem to stop growing.

```
1: 1253376 [1: 1253376] @ 0x409659 0x4097d7 0x40bb59 0x7b1fa7 0x7bac4b 0x7b8c3f 0x45b8d1
#	0x7b1fa6	github.com/gocql/gocql.(*Conn).exec+0x9a6		C:/Users/blah/src/github.com/gocql/gocql/conn.go:583
#	0x7bac4a	github.com/gocql/gocql.(*controlConn).writeFrame+0xba	C:/Users/blah/src/github.com/gocql/gocql/control.go:406
#	0x7b8c3e	github.com/gocql/gocql.(*controlConn).heartBeat+0x1ae	C:/Users/blah/src/github.com/gocql/gocql/control.go:76
```

**Comment by MichaelS11** [comment]
Still having memory leak issues. Looks like the goroutine writeFlusher sometimes is not being closed after doing a connection test. A connection test is creating a new connection, doing a simple query (select cql_version from system.local), then closing the connection. The goroutine writeFlusher is waiting on w.fcond.Wait(). 

https://github.com/gocql/gocql/blob/master/conn.go#L685

The only place I see a broadcast is here:

https://github.com/gocql/gocql/blob/master/conn.go#L653

I am still learning the code, but maybe a broadcast needs to be placed somewhere on quit, just in case?

Here is an example of one of the many goroutine writeFlusher getting stuck:
```
goroutine 76493 [semacquire, 651 minutes]:
sync.runtime_notifyListWait(0xc42272c710, 0x0)
/usr/local/go/src/runtime/sema.go:510 +0x10b
sync.(*Cond).Wait(0xc42272c700)
/usr/local/go/src/sync/cond.go:56 +0x80
github.com/gocql/gocql.(*writeCoalescer).writeFlusher(0xc4223a6780, 0xf4240, 0xc4223608a0)
/foo/go/src/github.com/gocql/gocql/conn.go:685 +0xc9
created by github.com/gocql/gocql.(*Session).dial
/foo/go/src/github.com/gocql/gocql/conn.go:273 +0xa6f
```

I have tried to find an exact test case to reproduce the problem not have not found one. It only seems to happen some of the time, not sure exactly what to trigger it.

**Comment by MichaelS11** [comment]
This seems related to https://github.com/gocql/gocql/pull/1198

**Comment by MichaelS11** [comment]
Looks to now be fixed

**Comment by alourie** [comment]
@MichaelS11 Thanks. I'm closing this now, can reopen if needed.

**Comment by MichaelS11** [comment]
@alourie 
This still seems to be an issue.

Build Oct 15, 2018 from master.

Seems to be stuck on `_, err = io.ReadFull(r, p[:1])`

https://github.com/gocql/gocql/blob/master/frame.go#L446


```
goroutine 435173 [IO wait, 598 minutes]:
internal/poll.runtime_pollWait(0x7fd401728300, 0x72, 0xc420f0aa48)
/usr/local/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc42051cd18, 0x72, 0xffffffffffffff00, 0xb0ee60, 0xe536c0)
/usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc42051cd18, 0xc420ed9000, 0x1000, 0x1000)
/usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc42051cd00, 0xc420ed9000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
/usr/local/go/src/internal/poll/fd_unix.go:157 +0x17d
net.(*netFD).Read(0xc42051cd00, 0xc420ed9000, 0x1000, 0x1000, 0xc4203d0380, 0xc420f0ad08, 0xc420f0ab90)
/usr/local/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc4200f2090, 0xc420ed9000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
/usr/local/go/src/net/net.go:176 +0x6a
bufio.(*Reader).Read(0xc420656660, 0xc420576648, 0x1, 0x9, 0x18, 0xc420f0ac10, 0x49b92d)
/usr/local/go/src/bufio/bufio.go:216 +0x238
io.ReadAtLeast(0xb0d080, 0xc420656660, 0xc420576648, 0x1, 0x9, 0x1, 0x40, 0x38, 0xa54040)
/usr/local/go/src/io/io.go:309 +0x86
io.ReadFull(0xb0d080, 0xc420656660, 0xc420576648, 0x1, 0x9, 0x72, 0x7fd4018d5f78, 0x0)
/usr/local/go/src/io/io.go:327 +0x58
github.com/gocql/gocql.readHeader(0xb0d080, 0xc420656660, 0xc420576648, 0x9, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
/blah/go/src/github.com/gocql/gocql/frame.go:446 +0xa1
github.com/gocql/gocql.(*Conn).recv(0xc420576600, 0x0, 0x0)
/blah/go/src/github.com/gocql/gocql/conn.go:507 +0xf1
github.com/gocql/gocql.(*Conn).serve(0xc420576600)
/blah/go/src/github.com/gocql/gocql/conn.go:471 +0x32
created by github.com/gocql/gocql.(*Session).dial
/blah/go/src/github.com/gocql/gocql/conn.go:245 +0x769
```

```
goroutine 435095 [select, 598 minutes]:
github.com/gocql/gocql.(*writeCoalescer).writeFlusher(0xc420a55da0, 0xf4240)
/blah/go/src/github.com/gocql/gocql/conn.go:739 +0x16e
created by github.com/gocql/gocql.newWriteCoalescer
/blah/go/src/github.com/gocql/gocql/conn.go:649 +0x132
```

When errors were received, there were: gocql: no hosts available in the pool

**Comment by MichaelS11** [comment]
Could someone please open this back up?

**Comment by MichaelS11** [comment]
Thanks for opening this back up.

What about changing `c.conn.SetReadDeadline(time.Time{})` to `c.conn.SetReadDeadline(time.Now().Add(c.timeout))` on line 504:
https://github.com/gocql/gocql/blob/master/conn.go#L504

Like line 262 is:
https://github.com/gocql/gocql/blob/master/conn.go#L262

**Comment by alourie** [comment]
@MichaelS11 makes sense to me.

**Comment by MichaelS11** [comment]
Cool, created PR https://github.com/gocql/gocql/pull/1234

**Comment by Zariel** [comment]
This is just going to cause connections to close themselves when they are idle and have the pool spin dialing them, that goroutine is intended to be blocked reading a header from cassandra for instance the control conn waiting for events.

The root cause of this is probably the control conn leaking because of timing issues.

**Comment by MichaelS11** [comment]
Ok, how is that solved?

**Comment by MichaelS11** [comment]
Can I get some help with this from someone who is familiar with the code? Give me a general idea on where the problem is and a general idea on how it should be solved?

**Comment by MichaelS11** [comment]
@alourie ? @Zariel ?

**Comment by beltran** [comment]
@MichaelS11 the problem is that it's not clear where the leak is happening. Before submitting a patch we should try to have proof of what's causing this. Are you still seeing this?

**Comment by Zariel** [comment]
The original issue from what I can tell was caused by the write coalescer not shutting down, which should now be fixed.

It is not immediately an issue if there is a goroutine sitting idle for long periods of time, though it does look suspect.

Is there a way to replicate this? Have you checked heap inuse objects where there are allocated?

**Comment by MichaelS11** [comment]
Yes, I am still seeing this. The stack trace is the one posted above, about 23 days ago from today.

The non-leak case is that a connection to the database is attempted, the network dial is done, then the network dial timeout is hit and the dial fails. The `func (c *Conn) recv() error {` is never called and so do not get stuck on the `_, err = io.ReadFull(r, p[:1])` line. All is well in the common case.

The issue is when the connection is tried, then the network dial works but Cassandra database has some kind of issue after that. That is when this called: https://github.com/gocql/gocql/blob/9de8c0414fd7ff72ee819f25a66e4942cb223873/conn.go#L498-L509 Then here: https://github.com/gocql/gocql/blob/9de8c0414fd7ff72ee819f25a66e4942cb223873/frame.go#L445-L446 Then the Goroutine can never die because the io.ReadFull never fails and the Cassandra database never responds.

I found this happens sometimes when a Cassandra database maintenance is performed and a connection test is done. A connection test is making sure the database is still accepting new connections by making a database connection then closing that connection. Easy to keep doing connection tests, say every 5 to 10 seconds to try to reproduce the issue. The problem is trying to get the Cassandra database in just the right fail state. The Cassandra database needs to respond to a network dial but not respond to the io.ReadFull.

**Comment by Zariel** [comment]
Do you have TCP heartbeats enabled? I guess we could move the control conn's heartbeating to the conn and use that to detect dead connections that aren't getting any writes.

Someone once told me that a slow node is worse than a down node, here we can't distinguish between a node being alive or not from the read side without sending frames and checking we get a response. We should be sending writes to this conn from the connection pool which would then trigger too many timeouts and close the connection.

**Comment by MichaelS11** [comment]
So I very noob to this driver code base, so apologies if I am off base. Would it be possible to enable SetReadDeadline and then handle the timeout errors as some kind of "temp error" in the calling functions? Isn't there some ground work for temp errors already?

**Comment by Zariel** [comment]
The read timeout is meaningless here because the protocol is asynchronous and bi directional. There would be no where to send this error beyond closing the connection which will cause the pool to redial it.

**Comment by Zariel** [comment]
There isn't anywhere for it to go up the stack, and the error will always be temporary so will always be retried.

**Comment by MichaelS11** [comment]
Which place does it not have anywhere to go up the stack?

Why does it always have be retried? Most database connection have a idle timeout. After some amount of time it is often wanted (at least in my option) to make a new connection rather than keep a idle connection around.

**Comment by MichaelS11** [comment]
@Zariel & @alourie 

Any thoughts on what is wanted? Happy to help if I can.

**Comment by zachaller** [comment]
We seem to be also getting hit by this, of note is I tried to pull #1245 to see if it helped hence why its running from my fork

```
goroutine 695 [IO wait, 21 minutes]:
internal/poll.runtime_pollWait(0x7fa2c1fe87d8, 0x72, 0xc0f4388aa8)
    /usr/local/go/src/runtime/netpoll.go:173 +0x66
internal/poll.(*pollDesc).wait(0xc000173598, 0x72, 0xffffffffffffff00, 0x1037da0, 0x18cac98)
    /usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0x9a
internal/poll.(*pollDesc).waitRead(0xc000173598, 0xc0f43c3000, 0x1000, 0x1000)
    /usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc000173580, 0xc0f43c3000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
    /usr/local/go/src/internal/poll/fd_unix.go:169 +0x179
net.(*netFD).Read(0xc000173580, 0xc0f43c3000, 0x1000, 0x1000, 0xc0f42a53a8, 0xc0f42a53e0, 0x44dbfa)
    /usr/local/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc00000e170, 0xc0f43c3000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
    /usr/local/go/src/net/net.go:177 +0x68
bufio.(*Reader).Read(0xc0f5027740, 0xc0f4e82048, 0x1, 0x9, 0xc000173580, 0x0, 0x20303d)
    /usr/local/go/src/bufio/bufio.go:216 +0x22f
io.ReadAtLeast(0x1035ee0, 0xc0f5027740, 0xc0f4e82048, 0x1, 0x9, 0x1, 0xeaf540, 0x0, 0x0)
    /usr/local/go/src/io/io.go:310 +0x88
io.ReadFull(0x1035ee0, 0xc0f5027740, 0xc0f4e82048, 0x1, 0x9, 0x0, 0x44e588, 0x29c25aa2534cb6)
    /usr/local/go/src/io/io.go:329 +0x58
github.com/zachaller/gocql.readHeader(0x1035ee0, 0xc0f5027740, 0xc0f4e82048, 0x9, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /root/go/src/github.com/zachaller/gocql/frame.go:446 +0xa1
github.com/zachaller/gocql.(*Conn).recv(0xc0f4e82000, 0x45eba1, 0xc0f47baa40)
    /root/go/src/github.com/zachaller/gocql/conn.go:520 +0xf2
github.com/zachaller/gocql.(*Conn).serve(0xc0f4e82000)
    /root/go/src/github.com/zachaller/gocql/conn.go:484 +0x31
created by github.com/zachaller/gocql.(*Session).dial
    /root/go/src/github.com/zachaller/gocql/conn.go:258 +0x764```

**Comment by MichaelS11** [comment]
Could we get some help fixing this?

**Comment by Zariel** [comment]
As far as I am concerned that goroutine is designed to block there and wait for frames, Cassandra is an asynchronous bidirectional protocol.

Take a look at pprof inuse memory and see what there is, is memory usage growing over time? Where is it being allocated?

**Comment by MichaelS11** [comment]
> I found this happens sometimes when a Cassandra database maintenance is performed and a connection test is done. A connection test is making sure the database is still accepting new connections by making a database connection then closing that connection. Easy to keep doing connection tests, say every 5 to 10 seconds to try to reproduce the issue. The problem is trying to get the Cassandra database in just the right fail state. The Cassandra database needs to respond to a network dial but not respond to the io.ReadFull.

**Comment by zachaller** [comment]
This seems to have fixed it for us, thanks for all the work!

**Comment by MichaelS11** [comment]
@Zariel Please reopen

PR https://github.com/gocql/gocql/pull/1258 seems to not close the heartBeat goroutine.

Example:

```
goroutine 34308 [select]:
github.com/gocql/gocql.(*Conn).heartBeat(0xc4232b2b00)
/build/go/src/github.com/gocql/gocql/conn.go:525 +0x160
created by github.com/gocql/gocql.(*Session).dial
/build/go/src/github.com/gocql/gocql/conn.go:259 +0x7c6
```

**Comment by MichaelS11** [comment]
@Zariel Would you like me to open a new issue instead?

**Comment by MichaelS11** [comment]
@alourie Can this be reopened? Or should I open a new issue?

**Comment by alourie** [comment]
@MichaelS11 let's open a new issue, and make it specific to heartBeats. I'll have a look at it next week.

When you open a ticket, please reference it here, so that we have the connection and validate that this issues is indeed handled.

Thanks!

**Comment by MichaelS11** [comment]
@alourie Thank you very much. Opened https://github.com/gocql/gocql/issues/1273
